torch/fx/experimental/symbolic_shapes.py

```diff
@@ -2293,6 +2381,95 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
                 self._add_guard(sympy.Eq(expr, concrete_val))  # type: ignore[arg-type]
         return concrete_val
 
+    def refine_ranges(self, guard: ShapeGuard) -> None:
+        def simplify(expr: sympy.Expr) -> sympy.Expr:
+            """
+            Simplification specialized for range refinement.
+            """
+            return self.maybe_simplify_floordiv(self.simplify(expr))
+
+        def simplify_until(expr: sympy.Expr, max_iterations: int = 10) -> sympy.Expr:
+            """
+            Calls 'simplify' either until it does not change or until it reaches the
+            maximum number of iterations.
+            """
+            max_iterations = 10
+            for _ in range(max_iterations):
+                previous, expr = expr, simplify(expr)
+                if expr == previous:
+                    break
+            return expr
+
+        RELOP_MIRROR = {
+            sympy.Ge: sympy.Le,
+            sympy.Gt: sympy.Lt,
+            sympy.Le: sympy.Ge,
+            sympy.Lt: sympy.Gt,
+        }
+
+        # List of expressions to be processed.
+        # Here, we try considering both LHS and RHS by mirroring the
+        # original expression: a < b ==> b > a
+        exprs = [guard.expr]
+
+        if type(guard.expr) in RELOP_MIRROR:
+            exprs.append(RELOP_MIRROR[type(guard.expr)](guard.expr.rhs, guard.expr.lhs))  # type: ignore[arg-type]
```

Comment:
Rather than doing things twice, why don't you simply flip the expression if needed to leave the free variable at the left?

Reply:
I guess that makes sense for univariate expressions. But, for multi-variate expressions, I still think we need this (e.g. `s0 < s1 // 2` -- not a real one). Should I move it to the next PR?

Reply:
I'm fine with it as is